src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.Windows.cs

```diff
@@ -0,0 +1,128 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>This class contains methods that are mainly used to manage native memory.</summary>
+    public static unsafe class NativeMemory
+    {
+        /// <summary>Allocates an aligned block of memory of the specified size and alignment, in bytes.</summary>
+        /// <param name="alignment">The alignment, in bytes, of the block to allocate.</param>
+        /// <param name="byteCount">The size, in bytes, of the block to allocate.</param>
+        /// <returns>A pointer to the allocated aligned block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="byteCount" /> of memory failed.</exception>
+        /// <remarks>
+        ///     <para>This method converts a <paramref name="byteCount" /> of <c>0</c> to <c>1</c>.</para>
+        ///     <para>This method is a thin wrapper over the C <c>aligned_alloc</c> API or a platform dependent aligned allocation API such as <c>_aligned_malloc</c> on Win32.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AlignedAlloc(nuint alignment, nuint byteCount)
+        {
+            nuint size = (byteCount != 0) ? byteCount : 1;
```

Comment:
The Windows implementation documents:
> If size is 0, malloc allocates a zero-length item in the heap and returns a valid pointer to that item.

Which seems to be different from giving a 1-byte allocation.

Reply:
Ok, I see that the aligned malloc calls the invalid arg handler for size == 0 in your other comment, so it makes sense to do something here.

The regular malloc should not need it on Windows.

Reply:
> Which seems to be different from giving a 1-byte allocation.

The malloc internally often returns chunk that is larger than what you asked for. I do not see the problem.

Reply:
The main concern with the 1-byte allocation is the risk for increased fragmentation across many allocations/frees.

Many small `1-byte` (or even `16-byte`, which is a common minimum on x64) seems to be "not good" and forcing the user to make the consideration, rather than defaulting to something for them for a conceivably "invalid" input, makes sense to me.

I don't feel strongly about it however and so I'm happy to do either. If we say its "invalid", then I think we'd just throw/treat it as OOM
If we say its valid, then we'd default to `size == 1` if the input was `0` and skip this step on platforms (like `Windows`) that explicitly document an acceptable corresponding behavior.

Reply:
Will default to `size == 1` as per https://github.com/dotnet/runtime/pull/54006#discussion_r649439220

Reply:
It is trade-off between convenience and squeezing the last cycle in every corner case.

I do not expect that well-written code wil hit the size == 0 frequently. If it does, it has a perf bug that one may want to fix.

You can make the same argument for `new byte[0]`. Having many 0-sized byte arrays is not good either, but we are not trying to actively prevent it by throwing argument exception for it since it would lead to need to special case it in user code in way too many places for little value.